packages/common/src/hooks/useHarmonyField.ts

```diff
@@ -0,0 +1,262 @@
+    validateFieldOnChange
+  } = props
+  const [field, meta, helpers] = useFormikField(props)
+  const { touched, error } = meta
+  const { setValue, setTouched } = helpers
+  const {
+    validateField,
+    validateOnBlur,
+    validateOnChange: validateFormOnChange,
+    submitCount
+  } = useFormikContext()
+
+  // Tracks whether the user is actively editing the field
+  const [isChanging, setIsChanging] = useState(false)
+
+  // Combine the form validateOnChange prop with this field's prop
+  const validateOnChange =
+    validateFieldOnChange === undefined
+      ? validateFormOnChange
+      : validateFieldOnChange
+
+  // Only show an error state if:
+  // - There's an error in validation, AND
+  // - The field has been touched (focused at least once), AND
+  // - The field can error before submission or the form has been submitted at least once, AND
+  // - The field is not currently being changed
+  const hasError = Boolean(
+    error &&
+      touched &&
+      (showErrorBeforeSubmit || submitCount > 0) &&
+      !isChanging
+  )
+
+  /**
+   * Sets the field to touched, validates it, and then sets isChanging to false.
+   * Essentially mimicking onBlur but ensuring validation finishes before
+   * allowing errors to show.
+   */
+  const validateAndSetIdle = useCallback(() => {
+    // We'll validate ourselves - pass false for validation
+    setTouched(true, false)
+    // Cast back to promise (the types are wrong)
+    const maybePromise = validateField(name) as Promise<any> | void
+    // Wait for validation before confirming we're not editing
+    if (maybePromise) {
+      maybePromise.then(() => {
+        setIsChanging(false)
+      })
+    } else {
+      setIsChanging(false)
+    }
+  }, [name, validateField, setIsChanging, setTouched])
+
+  // Debounced validateAndSetIdle for when the user stops typing for a bit
+  const debouncedValidateAndSetIdle = useDebouncedCallback(
+    validateAndSetIdle,
+    [validateAndSetIdle],
+    debouncedIdleMs
+  )
+
+  const onChangeText = useCallback(
+    async (e: React.ChangeEvent<any>) => {
+      // Add isChanging state so that errors don't show while typing.
+      setIsChanging(true)
+
+      // Extract value from event, taking care to get the nativeEvent.text
+      // for ReactNative.
+      let value = getEventTargetValue(e)
+
+      // Apply value transformations
+      if (transformValueOnChange) {
+        value = transformValueOnChange(value)
+      }
+
+      // All Formik's onChange handler does for text fields is call setValue.
+      // Calling setValue explicitly here instead of onChange for more control
+      // of the validation and so Formik doesn't have to infer the field name.
+      // Explicitly don't validate by setting the second param to false.
+      // Await the resulting promise to ensure the value is set before validating.
+      const maybePromise = setValue(value, false) as Promise<any> | void
+      if (maybePromise) {
```

Comment:
whats the scenario when this is a promise vs not?

Reply:
it's actually always a promise since we're not validating, but figure it's best to be a bit defensive anyway: https://github.com/jaredpalmer/formik/blob/0f960aaeeb0bdbef8312b5107cd3374884a0e62b/packages/formik/src/Formik.tsx#L583-L598

Reply:
surprisingly the types lie

Reply:
![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExMTFqY2FxOWU5OG1meWRmbzBtZDFuMW1nYW9hNXl0OGRxZXZrc2VxbyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/AssqAJR8ib5WmCNGOU/giphy.gif)

Reply:
Fair, thanks!